### PR TITLE
Add onSendBlock methods

### PR DESCRIPTION
**Pull request by robinmacharg**
## Goal

Add missing `add/remove/clearOnSendBlock()` methods to `Bugsnag` and `BugsnagConfiguration`.  These are used in this notifier in lieu of equivalent `onError` methods in others due to the tight requirements of code run in error handlers (e.g. no ObjC, async etc).

## Design

Straightforward exposure of access to the `BugsnagConfiguration` onSendBlocks array.

## Changeset

- `Bugsnag`, `BugsnagConfiguration` and associated tests.

- Some small non-functional rearrangement refactoring of headers has also been done.

## Tests

Unit tests.  Not tested prior to `start` being called.

### Outstanding Questions

- This pull request is ready for:
  - [ ] Initial review of the intended approach, not yet feature complete
  - [ ] Structural review of the classes, functions, and properties modified
  - [X] Final review

- [ ] The correct target branch has been selected (`master` for fixes, `next` for
  features)
- [ ] Consistency across platforms for structures or concepts added or modified
- [X] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [X] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: Source/BugsnagConfiguration.h [hunk 1]
```diff
@@ -61,8 +61,7 @@ typedef void (^BugsnagOnErrorBlock)(BugsnagEvent *_Nonnull report);
  *
  *  @return YES if the report should be sent
  */
-typedef bool (^BugsnagOnSendBlock)(NSDictionary *_Nonnull rawEventData,
-                                       BugsnagEvent *_Nonnull reports);
+typedef bool (^BugsnagOnSendBlock)(NSDictionary *_Nonnull rawEventData, BugsnagEvent *_Nonnull reports);
```
Comment (fractalwrench) [inline Source/BugsnagConfiguration.h]: I think it'd make sense to remove `rawEventData` and rename `reports` as `event`, as it could be a bit confusing having two parameters. In the past users have edited the `rawEventData` and been surprised that changes are not reflected in reports sent to the bugsnag dashboard.

File: Source/BugsnagConfiguration.m [hunk 2]
```diff
@@ -185,10 +185,27 @@ - (void)setUserMetadataFromUser:(BugsnagUser *)user {
     [self.metadata addAttribute:BSGKeyEmail withValue:user.emailAddress toTabWithName:BSGKeyUser];
 }
```
Comment (fractalwrench) [inline Source/BugsnagConfiguration.m]: The `onSendBlocks` property is publicly accessible. This (and the other callback arrays) should be hidden as part of the internal implementation and accessed via a class extension or similar mechanism.

File: Source/BugsnagConfiguration.m [hunk 3]
```diff
@@ -185,10 +185,27 @@ - (void)setUserMetadataFromUser:(BugsnagUser *)user {
...
+// MARK: - onSendBlock
+// =============================================================================
+
 - (void)addOnSendBlock:(BugsnagOnSendBlock)block {
     [(NSMutableArray *)self.onSendBlocks addObject:[block copy]];
 }
 
+- (void)removeOnSendBlock:(BugsnagOnSendBlock _Nonnull )block
+{
+    [(NSMutableArray *)self.onSendBlocks removeObject:block];
```
Comment (fractalwrench) [inline Source/BugsnagConfiguration.m]: This is more for my curiosity, but I'd be interested to know how this works, as `addOnSendBlock` makes a copy of the block whereas this is passed the original parameter. The test covering this scenario does seem to pass however.

File: Source/Bugsnag.h [hunk 4]
```diff
@@ -345,4 +352,28 @@ static NSString *_Nonnull const BugsnagSeverityInfo = @"info";
...
+ * Remove an onSend callback, if it exists
+ *
+ * @param block The block to remove
+ */
++ (void)removeOnSendBlock:(BugsnagOnSendBlock _Nonnull)block;
+
+/**
+ * Clear all onSend callbacks
+ */
++ (void)clearOnSendBlocks;
```
[the remark this answers is not in the document]
Comment (robinmacharg) [reply Source/Bugsnag.h]: Agreed, and yes, it does reduce the surface area, although part of me does like having the convenience methods around.

**Comment by robinmacharg** [comment]
> is this functionality defined anywhere in the notifier spec?

No, it's a divergence.  Due to the limitations of the environment during a crash (no async, or ObjC etc) we can't implement `onError` like the other notifiers.  `onSend` is the best we can do.  There is a separate task to document the divergence.

**Comment by fractalwrench** [comment]
>There is a separate task to document the divergence.

It would be preferable if the spec was up-to-date before we consider merging this PR.

Without a written specification that has been agreed upon, we run the risk of having to rework this interface or even scrap it entirely in favour of another solution. Documenting and scrutinising the requirements down beforehand reduces this risk.

File: Source/BugsnagConfiguration.h (hunk 1 above)
Comment (robinmacharg) [reply Source/BugsnagConfiguration.h]: Thanks for the background.  Happy to change it, if we're happy that nothing exclusively in the raw data could influence the return value.

**Comment by tomlongridge** [comment]
> Without a written specification that has been agreed upon, we run the risk of having to rework this interface or even scrap it entirely in favour of another solution. Documenting and scrutinising the requirements down beforehand reduces this risk.

I've raised an issue, but I'd seen this more of a rename to get it more like the spec until such time as we have resolved whether the concept of OnSend is a Cocoa deviation or something we'll introduce as a broader concept.

I'm happy to merge this once we have the add and remove methods (with no rawEventData parameter) but I'll kick off discussion on the issue to make sure it's covered before we release.

File: Source/BugsnagConfiguration.m (hunk 2 above)
Comment (robinmacharg) [reply Source/BugsnagConfiguration.m]: Agreed.  Done for `onSend` and `onSession`.

File: Source/BugsnagConfiguration.m (hunk 3 above)
Comment (robinmacharg) [reply Source/BugsnagConfiguration.m]: I’ve done a little digging; I think I understand it well enough to explain.  Failing that there’s a link at the end, and please correct me if I’m wrong.  Blocks are created initially on the stack, unlike other ObjC objects. The copy effectively _moves_ it to the heap, allowing it to be retained and used after the enclosing scope has exited.  The object pointer is updated.  So the usage in `remove()` will remove the correct - heap-stored - block from the array.  http://www.friday.com/bbum/2009/08/29/blocks-tips-tricks/
